sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc

```diff
@@ -0,0 +1,299 @@
+= sycl_ext_oneapi_bfloat16
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+// This is necessary for asciidoc, but not for asciidoctor
+:cpp: C++
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2022-2022 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 5.
+
+== Status
+
+This extension is implemented and fully supported by DPC++.
+
```

Comment:
There is still a restriction that needs to be documented.  Here is my proposed description:

```
[NOTE]
====
The DPC++ compiler has the following limitation when using this extension
in conjunction with ahead-of-time (AOT) compilation with the `-fsycl-targets`
compiler option.  When doing AOT compilation for an Intel GPU device via
`-fsycl-targets=spir64_gen -Xsycl-target-backend=spir64_gen "-device ..."`,
the compiler chooses either fallback or native support for bfloat16 according
to the device(s) specified in `...`.  Native support is used only if all of
these devices have native bfloat16 support.  As a result, AOT compiling for
multiple Intel GPU devices could result in the lower performance fallback
support even when running on a GPU that has native support.  Therefore, the
recommendation is to use AOT only when all Intel GPU devices have the same
type of bfloat16 support (all native support or all fallback support).

There is a similar limitation when AOT compiling for one Intel GPU device and
running on a different Intel GPU device.  In this case, the compiler chooses
either fallback or native bfloat16 support according to the device(s) specified
on the command line.  If you run the application on an Intel GPU device that
was not specified in the AOT compilation options, kernels will be just-in-time
(JIT) compiled for that device, but they will use the bfloat16 support (native
vs. fallback) that was selected by the compiler at AOT compilation time.  This
can result in lower performance (when AOT compiling for an old device and
running on a newer one) or in runtime errors (when AOT compiling for a new
device and running on an older one).  Therefore, when using AOT compilation for
Intel GPUs, we recommend running only on the same device(s) that are named in
the AOT compilation options.
====
```
```

Reply:
The first part of the note is correct and I have included it as-is. The second part is not quite right. Our runtime does not distinguish between specific GPU type, i.e., it does not distinguish between Gen9 and PVC. See my version of the note.

Reply:
What was wrong with the original wording?  Even though the SYCL runtime does not distinguish between the different GPU types, the device driver runtime does, and it is the device driver that JIT compiles from SPIR-V if there is no AOT version.  The wording you added misses the part about getting lower performance when running on a newer device if you compiled for an older one.  If you don't like my proposed wording, could you add something about that?  Maybe:

Old:

> If the fallback library had been chosen at AOT compilation time, then the binary will run on all Intel GPU devices.

New:

> If the fallback library was chosen at AOT compilation time, then the binary will run on all Intel GPU devices but you will not get the performance benefit of native support even when running on a new Intel GPU that has native support.

Reply:
An AOT binary built for an Intel GPU when run on any other Intel GPU will use the AOT code. It will not do JIT. The SYCL runtime distinguishes between GPU and CPU but not GPU type, i.e., whether it is Gen9 or PVC.

Reply:
This confused me for a while also, but I'm pretty sure I understand it now.  When DPC++ does AOT compilation, it invokes `ocloc` with the `-device` parameter specified on the DPC++ command line.  This results in a binary image that DPC++ treats as a "native binary" for GPU.  However, inside that native binary there are several things:

* One "second-level" native binary for each device listed in the `-device` command line option.
* A SPIR-V binary that was the original input to `ocloc`.

When the application runs, the DPC++ runtime sees the "native binary" for GPU and calls Level Zero to create a module from this native binary.  Level Zero then looks inside that native binary to see if there is a "second-level" native binary that matches the current GPU.  If there is, Level Zero uses it.  If there is not, Level Zero calls IGC to JIT compile an appropriate native binary from the internal SPIR-V.

Reply:
I was not aware of the second-level L0-triggered compilation. That cannot change the libraries supplied, so your wording is correct. I will change the Note.

Reply:
I have kept my version of the second part, but with the performance note in your last comment.

Reply:
Looks good.  Just a small comment on grammar above.